wandb/sdk/interface/interface_shared.py

```diff
@@ -0,0 +1,466 @@
+            raise Exception("Invalid record")
+        return record
+
+    @abstractmethod
+    def _publish(self, record: pb.Record, local: bool = None) -> None:
+        raise NotImplementedError
+
+    def _communicate(
+        self, rec: pb.Record, timeout: Optional[int] = 5, local: bool = None
+    ) -> Optional[pb.Result]:
+        return self._communicate_async(rec, local=local).get(timeout=timeout)
+
+    def _communicate_async(self, rec: pb.Record, local: bool = None) -> MessageFuture:
+        assert self._router
+        if self._process_check and self._process and not self._process.is_alive():
+            raise Exception("The wandb backend process has shutdown")
+        future = self._router.send_and_receive(rec, local=local)
+        return future
+
+    def communicate_login(
+        self, api_key: str = None, timeout: Optional[int] = 15
+    ) -> pb.LoginResponse:
+        login = self._make_login(api_key)
+        rec = self._make_request(login=login)
+        result = self._communicate(rec, timeout=timeout)
+        if result is None:
+            # TODO: friendlier error message here
+            raise wandb.Error(  # type: ignore[no-untyped-call]
+                "Couldn't communicate with backend after %s seconds" % timeout
+            )
+        login_response = result.response.login_response
+        assert login_response
+        return login_response
+
+    def _publish_defer(self, state: "pb.DeferRequest.DeferState.V") -> None:
+        defer = pb.DeferRequest(state=state)
+        rec = self._make_request(defer=defer)
+        self._publish(rec, local=True)
+
+    def publish_defer(self, state: int = 0) -> None:
+        self._publish_defer(cast("pb.DeferRequest.DeferState.V", state))
+
+    def _publish_header(self, header: pb.HeaderRecord) -> None:
+        rec = self._make_record(header=header)
+        self._publish(rec)
+
+    def publish_footer(self) -> None:
+        footer = pb.FooterRecord()
+        rec = self._make_record(footer=footer)
+        self._publish(rec)
+
+    def publish_final(self) -> None:
+        final = pb.FinalRecord()
+        rec = self._make_record(final=final)
+        self._publish(rec)
+
+    def publish_login(self, api_key: str = None) -> None:
+        login = self._make_login(api_key)
+        rec = self._make_request(login=login)
+        self._publish(rec)
+
+    def _publish_pause(self, pause: pb.PauseRequest) -> None:
+        rec = self._make_request(pause=pause)
+        self._publish(rec)
+
+    def _publish_resume(self, resume: pb.ResumeRequest) -> None:
+        rec = self._make_request(resume=resume)
+        self._publish(rec)
+
+    def _publish_run(self, run: pb.RunRecord) -> None:
+        rec = self._make_record(run=run)
+        self._publish(rec)
+
+    def _publish_config(self, cfg: pb.ConfigRecord) -> None:
+        rec = self._make_record(config=cfg)
+        self._publish(rec)
+
+    def publish_summary(self, summary_record: sr.SummaryRecord) -> None:
```

Comment:
I see it's already implemented in `InterfaceBase`, why do we need it here as well?

Reply:
oops.  fixed.